include/helib/query.h

```diff
@@ -515,6 +603,126 @@ class QueryBuilder
 
     return std::move(convertStack.top());
   }
+
+  /**
+   * @brief Given a vecvec where outer clauses correspond to ANDs and inner
+   * clauses correspond to ORs, deletes duplicated columns and columns for which
+   * both positive and negation occur from inner clauses, and deletes empty
+   * clauses, inplace.
+   * @param expr a vecvec corresponding to CNF.
+   */
+  void tidy(vecvec& expr) const
+  {
+    vecvec x;
+    for (auto& y : expr) {
+      tidyClause(y);
+      if (y.size() != 0)
+        x.push_back(y);
+    }
+    expr = x;
+  }
+
+  /**
+   * @brief build the weights for a query given as an AND of ORs.
+   * @param expr inner groups correspond to ORs, and we take the AND across
+   * all inner groups. (i + 1) corresponds to column i, and negatives to
+   * negations.
+   * @param columns The number of columns in the database.
+   * @return QueryType with the correct weights.
+   */
+  QueryType buildWeights(const vecvec& expr, const long columns) const
+  {
+    assertTrue(expr.size() > 0, "Cannot build weights for an empty query.");
+    bool containsOR = false;
+
+    vecvec Fs(expr.size());
+    {
+      std::vector<long> v(columns);
+      std::iota(v.begin(), v.end(), 0);
+      std::fill(Fs.begin(), Fs.end(), v);
+    }
+    std::vector<long> mus(expr.size(), 1);
+    std::vector<Matrix<long>> taus;
+    taus.reserve(expr.size());
+
+    // Create the taus
+    for (size_t i = 0; i < expr.size(); ++i) { // Each tau
+      assertTrue(expr[i].size() > 0,
+                 "empty clause found at index " + std::to_string(i) +
+                     ", function called without Tidy().");
+      mus[i] = 0;                 // Set mu to zero.
+      Matrix<long> M(columns, 1); // Create temp tau matrix
+      containsOR = (expr[i].size() > 1) ? true : containsOR;
+      for (size_t j = 0; j < expr[i].size(); ++j) // Each column index
+      {
+        // column should not already have been seen in this clause
+        assertTrue(M(std::abs(expr[i][j]) - 1, 0) == 0,
+                   "Column " + std::to_string(std::abs(expr[i][j]) - 1) +
+                       " repeated in clause " + std::to_string(i) +
+                       ", function called without Tidy().");
+        // add one to the offset for each NOT
+        if (expr[i][j] < 0)
+          mus[i] += 1;
+        // mark NOT columns as -1, columns as 1
+        M(std::abs(expr[i][j]) - 1, 0) = ((expr[i][j] >= 0) ? 1 : -1);
+      }
+      taus.push_back(std::move(M));
+    }
+    return QueryType(std::move(Fs),
+                     std::move(mus),
+                     std::move(taus),
+                     containsOR);
+  }
+  /**
+   * @brief Given a vecvec intepreted as an AND of ORs, return a vecvec
+   * corresponding to the negation. Called by expandOr()
+   */
```

Comment:
Missing `@param clause` and `@return`.

Reply:
I've expanded this here https://github.com/helibproject/HElib/blob/14c3b9f9f57865c09ae340b7322bdfceb16caa64/include/helib/query.h#L678